datafusion/expr/src/binary_rule.rs

```diff
@@ -801,4 +804,230 @@ mod tests {
+            Operator::Divide,
+            DataType::Decimal128(23, 11)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Modulo,
+            DataType::Decimal128(10, 2)
+        );
+        // TODO add other data type
+        Ok(())
+    }
+
+    #[test]
+    fn test_type_coercion_compare() -> Result<()> {
+        // boolean
+        test_coercion_binary_rule!(
+            DataType::Boolean,
+            DataType::Boolean,
+            Operator::Eq,
+            DataType::Boolean
+        );
+        // float
+        test_coercion_binary_rule!(
+            DataType::Float32,
+            DataType::Int64,
+            Operator::Eq,
+            DataType::Float32
+        );
+        test_coercion_binary_rule!(
+            DataType::Float32,
+            DataType::Float64,
+            Operator::GtEq,
+            DataType::Float64
+        );
+        // signed integer
+        test_coercion_binary_rule!(
+            DataType::Int8,
+            DataType::Int32,
+            Operator::LtEq,
+            DataType::Int32
+        );
+        test_coercion_binary_rule!(
+            DataType::Int64,
+            DataType::Int32,
+            Operator::LtEq,
+            DataType::Int64
+        );
+        // unsigned integer
+        test_coercion_binary_rule!(
+            DataType::UInt32,
+            DataType::UInt8,
+            Operator::Gt,
+            DataType::UInt32
+        );
+        // numeric/decimal
+        test_coercion_binary_rule!(
+            DataType::Int64,
+            DataType::Decimal128(10, 0),
+            Operator::Eq,
+            DataType::Decimal128(20, 0)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int64,
+            DataType::Decimal128(10, 2),
+            Operator::Lt,
+            DataType::Decimal128(22, 2)
+        );
+        test_coercion_binary_rule!(
+            DataType::Float64,
+            DataType::Decimal128(10, 3),
+            Operator::Gt,
+            DataType::Decimal128(30, 15)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int64,
+            DataType::Decimal128(10, 0),
+            Operator::Eq,
+            DataType::Decimal128(20, 0)
+        );
+        test_coercion_binary_rule!(
+            DataType::Decimal128(14, 2),
+            DataType::Decimal128(10, 3),
+            Operator::GtEq,
+            DataType::Decimal128(15, 3)
+        );
+
+        // TODO add other data type
+        Ok(())
+    }
+
+    #[test]
+    fn test_type_coercion_logical_op() -> Result<()> {
+        test_coercion_binary_rule!(
+            DataType::Boolean,
+            DataType::Boolean,
+            Operator::And,
+            DataType::Boolean
+        );
+
+        test_coercion_binary_rule!(
+            DataType::Boolean,
+            DataType::Boolean,
+            Operator::Or,
+            DataType::Boolean
```

Comment:
Is there any other types that are coerced to boolean for logical operations? Or are the tests for boolean just showing that boolean types are not changed when coerced?